ParallelStreamsLoadTest OOM Java heap space

https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.system_aarch64_linux_Nightly_testList_3/17 - [ubu22-aarch64-4](https://openj9-jenkins.osuosl.org/computer/ubu22-aarch64-4)
ParallelStreamsLoadTest_J9_2 `-Xjit -Xgcpolicy:balanced -Xnocompressedrefs `

https://openj9-artifactory.osuosl.org/artifactory/ci-openj9/Test/Test_openjdk8_j9_sanity.system_aarch64_linux_Nightly_testList_3/17/system_test_output.tar.gz

```
00:02:31  LT  Caused by: java.lang.OutOfMemoryError: Java heap space
00:02:31  LT  	at java.util.HashMap.resize(HashMap.java:705)
00:02:31  LT  	at java.util.HashMap.merge(HashMap.java:1236)
00:02:31  LT  	at java.util.stream.Collectors.lambda$mapMerger$7(Collectors.java:320)
00:02:31  LT  	at java.util.stream.ReduceOps$3ReducingSink.combine(ReduceOps.java:174)
00:02:31  LT  	at java.util.stream.ReduceOps$3ReducingSink.combine(ReduceOps.java:160)
00:02:31  LT  	at java.util.stream.ReduceOps$ReduceTask.onCompletion(ReduceOps.java:754)
00:02:31  LT  	at java.util.concurrent.CountedCompleter.tryComplete(CountedCompleter.java:577)
00:02:31  LT  	at java.util.stream.AbstractTask.compute(AbstractTask.java:328)
00:02:31  LT  	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
00:02:31  LT  	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
```

```
1STHEAPTYPE    Object Memory
NULL           id                 start              end                size               space
1STHEAPSPACE   0x0000FFFF8C0B95C0 0x0000FFFF4C000000 0x0000FFFF8C000000 0x0000000040000000 Flat
1STHEAPEXT     0x0000FFFF8C0632C0 0x0000FFFDEC000000 0x0000FFFF4BF00000 0x000000015FF00000 VirtualLargeObjectHeap (off-heap)
NULL
1STHEAPTOTAL   Total memory:                    1073741824 (0x0000000040000000)
1STHEAPINUSE   Total memory in use:              520733160 (0x000000001F09C1E8)
2STHEAPSPCUSE  Flat memory in use:               467255784 (0x000000001BD9C1E8)
2STHEAPEXTUSE  Off-heap memory in use:            53477376 (0x0000000003300000)
1STHEAPFREE    Total memory free:                553008664 (0x0000000020F63E18)

1STGCHTYPE     GC History
3STHSTTYPE     04:02:14:190080522 GMT j9mm.468 -   Cycle End: type 3 approximateFreeMemorySize 566231040
3STHSTTYPE     04:02:14:190002203 GMT j9mm.346 -   PGC end: workstackoverflow=0 overflowcount=0 heapFree=566231040/1073741824
3STHSTTYPE     04:02:14:186726059 GMT j9mm.90 -   GlobalGC collect complete
3STHSTTYPE     04:02:14:186678380 GMT j9mm.137 -   Compact end: bytesmoved=0
3STHSTTYPE     04:02:14:182452343 GMT j9mm.136 -   Compact start: reason=no compaction
3STHSTTYPE     04:02:14:182365386 GMT j9mm.94 -   Class unloading end: classloadersunloaded=0 classesunloaded=0
3STHSTTYPE     04:02:14:182327266 GMT j9mm.60 -   Class unloading start
3STHSTTYPE     04:02:14:182325746 GMT j9mm.466 -   CopyForward end
3STHSTTYPE     04:02:14:117673428 GMT j9mm.465 -   CopyForward start
3STHSTTYPE     04:02:14:112524297 GMT j9mm.345 -   PGC start: globalcount=152 nextGMPIncrement=0
3STHSTTYPE     04:02:14:112523817 GMT j9mm.467 -   Cycle Start: type 3 approximateFreeMemorySize 0
3STHSTTYPE     04:02:14:112306063 GMT j9mm.101 -   J9AllocateIndexableObject() returning NULL! 2097176 bytes requested for object of class 0000FFFF8C2CCD00 from memory space 'Flat' id=0000FFFF8C0B95C0
3STHSTTYPE     04:02:14:112218706 GMT j9mm.134 -   Allocation failure end: newspace=0/0 oldspace=0/507510784 loa=0/0
3STHSTTYPE     04:02:14:112192747 GMT j9mm.470 -   Allocation failure cycle end: newspace=0/0 oldspace=0/507510784 loa=0/0
3STHSTTYPE     04:02:14:112060470 GMT j9mm.468 -   Cycle End: type 5 approximateFreeMemorySize 0
3STHSTTYPE     04:02:14:111986153 GMT j9mm.475 -   GlobalGC end: workstackoverflow=0 overflowcount=0 memory=0/507510784
3STHSTTYPE     04:02:14:110084888 GMT j9mm.90 -   GlobalGC collect complete
```

--- a/runtime/gc_vlhgc/IncrementalGenerationalGC.cpp
+++ b/runtime/gc_vlhgc/IncrementalGenerationalGC.cpp
@@ -1021,6 +1021,8 @@ MM_IncrementalGenerationalGC::partialGarbageCollectPostWork(MM_EnvironmentVLHGC 
 	PORT_ACCESS_FROM_ENVIRONMENT(env);
 	env->_cycleState->_endTime = j9time_hires_clock();
 
+	verifyHeapSizing(env, false);
+
 	reportGCCycleFinalIncrementEnding(env);
 	reportGCIncrementEnd(env);
 	reportPGCEnd(env);
@@ -1257,6 +1259,8 @@ MM_IncrementalGenerationalGC::runGlobalGarbageCollection(MM_EnvironmentVLHGC *en
 	PORT_ACCESS_FROM_ENVIRONMENT(env);
 	env->_cycleState->_endTime = j9time_hires_clock();
 
+	verifyHeapSizing(env, true);
+
 	reportGCCycleFinalIncrementEnding(env);
 	/* TODO: TEMPORARY: This is a temporary call that should be deleted once the new verbose format is in place */
 	/* NOTE: May want to move any tracepoints up into this routine */
@@ -1339,7 +1343,13 @@ MM_IncrementalGenerationalGC::preProcessPGCUsingCopyForward(MM_EnvironmentVLHGC 
 	 * NOTE: A second estimate for free tenure is later made - The lowest estimate for free tenure is used in heap sizing calculations
 	 */
 	_extensions->globalVLHGCStats._heapSizingData.freeTenure = freeMemoryForSurvivor;
-
+	/* in some case free eden size is significant, should not be counted as freeTenure */
+	uintptr_t allocatedSinceLastPGC = getAllocatedSinceLastPGC();
+	uintptr_t edenSize = getCurrentEdenSizeInBytes(NULL);
+	if (edenSize >= allocatedSinceLastPGC) {
+		Assert_MM_true(_extensions->globalVLHGCStats._heapSizingData.freeTenure >= (edenSize - allocatedSinceLastPGC));
+		_extensions->globalVLHGCStats._heapSizingData.freeTenure -= edenSize - allocatedSinceLastPGC;
+	}
 	cycleState->_vlhgcIncrementStats._copyForwardStats._freeMemoryBefore = freeMemoryForSurvivor;
 	cycleState->_vlhgcIncrementStats._copyForwardStats._totalMemoryBefore = _extensions->getHeap()->getMemorySize();
 
@@ -2610,4 +2620,15 @@ MM_IncrementalGenerationalGC::getBytesScannedInGlobalMarkPhase()
 		bytesScanned = _persistentGlobalMarkPhaseState._vlhgcCycleStats._markStats._bytesScanned;
 	}
 	return bytesScanned;
+}
+
+void
+MM_IncrementalGenerationalGC::verifyHeapSizing(MM_EnvironmentVLHGC *env, bool isGlobalGC)
+{
+	Assert_GC_true_with_message(env, (((MM_GlobalAllocationManagerTarok *)_extensions->globalAllocationManager)->getFreeRegionCount() >= _schedulingDelegate.getCurrentEdenSizeInRegions(env)),
+				"verifyHeapSizing freeRegionCount(%zu) is less than EdenRegionCount(%zu), allocatedSinceLastPGC=%zu bytes in current %s\n",
+				((MM_GlobalAllocationManagerTarok *)_extensions->globalAllocationManager)->getFreeRegionCount(),
+				_schedulingDelegate.getCurrentEdenSizeInRegions(env),
+				getAllocatedSinceLastPGC(),
+				(isGlobalGC)?"Global GC":"PGC");
 }

--- a/runtime/gc_vlhgc/IncrementalGenerationalGC.hpp
+++ b/runtime/gc_vlhgc/IncrementalGenerationalGC.hpp
@@ -147,6 +147,12 @@ private:
 	 * Tooling report the end of an unload phase.
 	 */
 	void reportClassUnloadingEnd(MM_EnvironmentBase *env);
+
+	/**
+	 * Confirm free region count is equal or more than eden region count.
+	 */
+	void verifyHeapSizing(MM_EnvironmentVLHGC *env, bool isGlobalGC);
+
 #endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
 
 protected:
@@ -413,6 +419,11 @@ public:
 	MMINLINE UDATA getCurrentEdenSizeInBytes(MM_EnvironmentVLHGC *env)
 	{
 		return _schedulingDelegate.getCurrentEdenSizeInBytes(env);
+	}
+
+	MMINLINE void recalculateEdenSize(MM_EnvironmentVLHGC *env)
+	{
+		_schedulingDelegate.recalculateEdenSize(env);
 	}
 
 	MM_IncrementalGenerationalGC(MM_EnvironmentVLHGC *env, MM_HeapRegionManager *manager);

--- a/runtime/gc_vlhgc/MemorySubSpaceTarok.cpp
+++ b/runtime/gc_vlhgc/MemorySubSpaceTarok.cpp
@@ -877,7 +877,7 @@ MM_MemorySubSpaceTarok::collectorExpand(MM_EnvironmentBase *env)
 	Assert_MM_true((0 == expandSize) || (_heapRegionManager->getRegionSize() == expandSize));
 
 	_extensions->heap->getResizeStats()->setLastExpandReason(SATISFY_COLLECTOR);
-	
+
 	/* expand by a single region */
 	/* for the most part the code path is not multi-threaded safe, so we do this under expandLock */
 	uintptr_t expansionAmount= expand(env, expandSize);
@@ -942,8 +942,21 @@ MM_MemorySubSpaceTarok::performResize(MM_EnvironmentBase *env, MM_AllocateDescri
 		resizeAmount = -(intptr_t)performContract(env, allocDescription);
 	} else if (_expansionSize != 0) {
 		resizeAmount = performExpand(env);
+	} else {
+		/**
+		 * In case there is no heap resize, check if there is the case that free size is smaller than eden size
+		 * due to the conflict between eden resize and heap resize, recalculateEdenSize if it happens.
+		 */
+		uintptr_t freeBytes = _globalAllocationManagerTarok->getFreeRegionCount()*_heapRegionManager->getRegionSize();
+		MM_IncrementalGenerationalGC *collector = (MM_IncrementalGenerationalGC*)_extensions->getGlobalCollector();
+		uintptr_t edenSizeInBytes = collector->getCurrentEdenSizeInBytes((MM_EnvironmentVLHGC *)env);
+		if (edenSizeInBytes > freeBytes) {
+			collector->recalculateEdenSize((MM_EnvironmentVLHGC *)env);
+			edenSizeInBytes = collector->getCurrentEdenSizeInBytes((MM_EnvironmentVLHGC *)env);
+		}
+		Assert_MM_true(freeBytes >= edenSizeInBytes);
 	}
-	
+
 	env->popVMstate(oldVMState);
 
 	return resizeAmount;

--- a/runtime/gc_vlhgc/SchedulingDelegate.cpp
+++ b/runtime/gc_vlhgc/SchedulingDelegate.cpp
@@ -419,7 +419,7 @@ MM_SchedulingDelegate::partialGarbageCollectCompleted(MM_EnvironmentVLHGC *env, 
 
 	/* Check eden size based off of new PGC stats */
 	checkEdenSizeAfterPgc(env, globalSweepHappened);
-	calculateEdenSize(env);
+	calculateEdenSize(env, true);
 	/* Recalculate GMP intermission after (possibly) resizing eden */
 	calculateAutomaticGMPIntermission(env);
 	estimateMacroDefragmentationWork(env);
@@ -1343,7 +1343,7 @@ MM_SchedulingDelegate::updateSurvivalRatesAfterCopyForward(double thisEdenSurviv
 }
 
 void 
-MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
+MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env, bool allowTotalHeapResize)
 {
 	uintptr_t regionSize = _regionManager->getRegionSize();
 	uintptr_t previousEdenSize = _edenRegionCount * regionSize;
@@ -1362,7 +1362,7 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
 	Assert_MM_true(edenMaximumCount >= 1);
 	Assert_MM_true(edenMaximumCount >= edenMinimumCount);
 
-	/* Allow eden to expand as much as it wants, as long as the total heap can expand to accomodate it */
+	/* Allow eden to expand as much as it wants, as long as the total heap can expand to accommodate it. */
 	uintptr_t desiredEdenCount = OMR_MAX(edenMaximumCount, edenMinimumCount);
 	intptr_t desiredEdenChangeSize = (intptr_t)desiredEdenCount - (intptr_t)_edenRegionCount;
 	/* Determine if eden should try to grow anyways, or if the heap is tight on memory */
@@ -1372,41 +1372,41 @@ MM_SchedulingDelegate::calculateEdenSize(MM_EnvironmentVLHGC *env)
 
 	Trc_MM_SchedulingDelegate_calculateEdenSize_dynamic(env->getLanguageVMThread(), desiredEdenCount, _edenSurvivalRateCopyForward, _nonEdenSurvivalCountCopyForward, freeRegions, edenMinimumCount, edenMaximumCount);
 
-	/* Make sure that the total heap can expand enough to satisfy the desired change in eden size
-	 * If heap is fully expanded (or close to) make sure that there are enough free regions to satisfy given eden size change
-	 */
-	intptr_t maxEdenChange = 0;
-	uintptr_t maxEdenRegionCount = _regionManager->getTableRegionCount();
-	bool edenIsVerySmall = (_edenRegionCount * 64) < maxEdenRegionCount;
+	/* Eden size can not be bigger than free region size. */
+	intptr_t maxEdenChange = freeRegions - _edenRegionCount;
+	if (allowTotalHeapResize) {
 
-	/* Eden will be stealing free regions from the entire heap, without telling the heap to grow.
-	 * Note: the eden sizing logic knows how much free memory is available in the heap, and knows to not grow too much.
-	 * eden size can not be bigger than free region size.
-	 */
-	maxEdenChange = freeRegions - _edenRegionCount;
-
-	if (0 == maxHeapExpansionRegions) {
-		_extensions->globalVLHGCStats._heapSizingData.edenRegionChange = 0;
-	} else {
-		/* Eden will inform the total heap resizing logic, that it needs to change total heap size in order to maintain same "tenure" size */
 		maxEdenChange += maxHeapExpansionRegions;
+		/* Make sure that the total heap can expand enough to satisfy the desired change in eden size.
+		 * If heap is fully expanded (or close to) make sure that there are enough free regions to satisfy given eden size change.
+		 */
 		intptr_t edenChangeWithSurvivorHeadroom = desiredEdenChangeSize;
-
-		/* Total heap needs to be aware that by changing eden size, the amount of survivor space might also need to change */
+		/* Proportionally adjust survivor area. */
 		if (0 < desiredEdenChangeSize) {
-			edenChangeWithSurvivorHeadroom = desiredEdenChangeSize + (intptr_t)ceil(((double)desiredEdenChangeSize * _edenSurvivalRateCopyForward));
-		} else if ((0 > desiredEdenChangeSize) && !edenIsVerySmall) {
-			/* If eden is shrinking, only factor adjusting in survivor regions for total heap resizing when eden is not very small.
-			 * Factoring in survivor regions when eden is tiny can lead to some innacuracies, and reduce free non-eden regions, which may impact performance
+			/* Eden contraction must not trigger non-eden contraction. Doing so could
+			 * incorrectly result in zero free memory even when the heap is still expandable.
 			 */
-			edenChangeWithSurvivorHeadroom = desiredEdenChangeSize + (intptr_t)floor(((double)desiredEdenChangeSize * _edenSurvivalRateCopyForward));
+			edenChangeWithSurvivorHeadroom += (intptr_t)ceil((double)desiredEdenChangeSize * _edenSurvivalRateCopyForward);
 		}
-		_extensions->globalVLHGCStats._heapSizingData.edenRegionChange = OMR_MIN(maxEdenChange, edenChangeWithSurvivorHeadroom);
+		/* Inform the total heap resizing logic, that it needs to change total heap size in order to maintain same "non-eden" size. */
+		if (freeRegions > _edenRegionCount) {
+			_extensions->globalVLHGCStats._heapSizingData.edenRegionChange = OMR_MIN(maxHeapExpansionRegions, edenChangeWithSurvivorHeadroom);
+		} else {
+			/* PGC has not recovered enough regions to accommodate even for the current eden size.
+			 * So, let's expand heap by that deficit (capped to the maximum that heap expansion allows), plus whatever the new eden size requires).
+			 */
+			_extensions->globalVLHGCStats._heapSizingData.edenRegionChange = OMR_MIN(maxHeapExpansionRegions, edenChangeWithSurvivorHeadroom + (intptr_t)(_edenRegionCount - freeRegions));
+		}
 	}
 
+	/**
+	 * We account for maxEdenChange late, since it could have changed based on allowTotalHeapResize value. Meanwhile,
+	 * we had notification for full heap resizing based on old maxEdenChange value, but that is ok, since the notification accounted
+	 * for maxHeapExpansionRegions, what is exactly by how much maxEdenChange is potentially changed.
+	 */
 	desiredEdenChangeSize = OMR_MIN(maxEdenChange, desiredEdenChangeSize);
-
-	_edenRegionCount = (uintptr_t)OMR_MAX(1, ((intptr_t)_edenRegionCount + desiredEdenChangeSize));
+	Assert_MM_true(0 <= ((intptr_t)_edenRegionCount + desiredEdenChangeSize));
+	_edenRegionCount = _edenRegionCount + desiredEdenChangeSize;
 
 	Trc_MM_SchedulingDelegate_calculateEdenSize_Exit(env->getLanguageVMThread(), (_edenRegionCount * regionSize));
 }

--- a/runtime/gc_vlhgc/SchedulingDelegate.hpp
+++ b/runtime/gc_vlhgc/SchedulingDelegate.hpp
@@ -271,10 +271,11 @@ private:
 	/**
 	 * Following a GC, recalculate the Eden size for the next PGC.
 	 * This is typically the same as GCExtensions->tarokeEdenSize, but may be smaller if
-	 * insufficient memory is available
+	 * insufficient memory is available.
 	 * @param env[in] the main GC thread
+	 * @param allowTotalHeapResize[in] if true, allow total heap to resize to accommodate for eden resizing (typically after a PGC)
 	 */
-	void calculateEdenSize(MM_EnvironmentVLHGC *env);
+	void calculateEdenSize(MM_EnvironmentVLHGC *env, bool allowTotalHeapResize = false);
 
 	/**
 	 * Compute what the ideal eden size should be, and return by how many regions eden should change
@@ -658,6 +659,14 @@ public:
 	 * Adjust internal structures to reflect the change in heap size.
 	 */
 	void heapReconfigured(MM_EnvironmentVLHGC *env);
+
+	/**
+	 * call calculateEdenSize() without allowTotalHeapResize option
+	 */
+	void recalculateEdenSize(MM_EnvironmentVLHGC *env)
+	{
+		calculateEdenSize(env);
+	}
 
 	double getAvgEdenSurvivalRateCopyForward(MM_EnvironmentVLHGC *env) { return _edenSurvivalRateCopyForward; }
 

